### PR TITLE
CVS-103854 Fix KFS REST & tfs/kfs gRPC

### PR DESCRIPTION
**Pull request by dkalinowski**
- fix issue with negative value REST shape causing abnormal gRPC proto creation
- fix issue with gRPC negative shape value, causing abnormal behavior in post-validation steps
- added unit tests
- added support for signed values in shape in test framework for better corner case testing

### CODE REVIEW

**Review by dtrawins — APPROVED** [review]

**Review by rasapala — APPROVED** [review]
LGTM

File: src/rest_parser.cpp [hunk 1]
```diff
@@ -618,6 +618,9 @@ Status KFSRestParser::parseInput(rapidjson::Value& node, bool onlyOneInput) {
         if (!dim.IsInt()) {
             return StatusCode::REST_COULD_NOT_PARSE_INPUT;
         }
+        if (dim.GetInt() <= 0) {
```
Comment (atobiszei) [inline src/rest_parser.cpp:621]: Please add check in model instance before model reload to check status from request validation.
We should only extract request shape & batch size only if validation returns: StatusCode::OK, reshape_requred, batchsize_change_required

**Review by atobiszei — CHANGES_REQUESTED** [review]
Improve ModelInstance::infer()

File: src/modelinstance.cpp [hunk 2]
```diff
@@ -1154,9 +1154,11 @@ Status ModelInstance::infer(const RequestType* requestProto,
     if (!status.ok())
         return status;
     status = validate(requestProto);
-    auto requestBatchSize = getRequestBatchSize(requestProto, this->getBatchSizeIndex());
-    auto requestShapes = getRequestShapes(requestProto);
-    status = reloadModelIfRequired(status, requestBatchSize, requestShapes, modelUnloadGuardPtr);
+    if (status.batchSizeChangeRequired() || status.reshapeRequired()) {
```
Comment (dkalinowski) [inline src/modelinstance.cpp:1157]: needs test for shape=auto and sending -5

File: src/prediction_service_utils.cpp [hunk 3]
```diff
@@ -82,7 +87,12 @@ std::optional<Dimension> getRequestBatchSize(const tensorflow::serving::PredictR
         SPDLOG_DEBUG("Failed to get batch size of a request. Batch size index out of shape range. Validation of request failed");
         return std::nullopt;
     }
-    return Dimension(requestInput.tensor_shape().dim(batchSizeIndex).size());
+    auto value = requestInput.tensor_shape().dim(batchSizeIndex).size();
```
Comment (dkalinowski) [inline src/prediction_service_utils.cpp:90]: needs tests for those utils

**Comment by dkalinowski** [comment]
needs fix with c-api unsigned to signed cast

File: src/modelinstance.cpp (hunk 2 above)
Comment (dkalinowski) [reply src/modelinstance.cpp:1157]: done

File: src/rest_parser.cpp (hunk 1 above)
Comment (dkalinowski) [reply src/rest_parser.cpp:621]: done

File: src/prediction_service_utils.cpp (hunk 3 above)
Comment (dkalinowski) [reply src/prediction_service_utils.cpp]: removed value check since the contract is to use this function only after passed validation

File: src/test/test_utils.hpp [hunk 4]
```diff
@@ -154,24 +154,24 @@ ovms::tensor_map_t prepareTensors(
     const std::unordered_map<std::string, ovms::Shape>&& tensors,
     ovms::Precision precision = ovms::Precision::FP32);
 
-void preparePredictRequest(tensorflow::serving::PredictRequest& request, inputs_info_t requestInputs, const std::vector<float>& data = {});
+void preparePredictRequest(tensorflow::serving::PredictRequest& request, inputs_info_t requestInputs, const std::vector<float>& data = {}, std::optional<int64_t> overrideBatch = std::nullopt);
```
Comment (atobiszei) [inline src/test/test_utils.hpp:157]: Wouldn't it be better to replace:
```test/test_utils.hpp:48:using inputs_info_t = std::map<std::string, std::tuple<ovms::shape_t, ovms::Precision>>;```
with
```test/test_utils.hpp:48:using inputs_info_t = std::map<std::string, std::tuple<std::vector<int64_t>, ovms::Precision>>;```

Comment (atobiszei) [reply src/test/test_utils.hpp:157]: Beside that LGTM

**Review by atobiszei — APPROVED** [review]

File: src/test/test_utils.hpp (hunk 4 above)
Comment (dkalinowski) [reply src/test/test_utils.hpp]: done

**Review by atobiszei — APPROVED** [review]
